nifi-nar-bundles/nifi-graph-bundle/nifi-neo4j-3x-cypher-service/src/main/java/org/apache/nifi/graph/Neo4JCypher3xClientService.java

```diff
@@ -0,0 +1,318 @@
+                configBuilder.toConfig());
+    }
+
+    /**
+     * Helper method to help testability
+     * @return Driver instance
+     */
+    protected Driver getNeo4JDriver() {
+        return neo4JDriver;
+    }
+
+    @OnEnabled
+    public void onEnabled(final ConfigurationContext context) {
+        try {
+            neo4JDriver = getDriver(context);
+        } catch(Exception e) {
+            getLogger().error("Error while getting connection " + e.getLocalizedMessage(),e);
+            throw new ProcessException("Error while getting connection" + e.getLocalizedMessage(),e);
+        }
+        getLogger().info("Neo4JCypherExecutor connection created for url {}",
+                new Object[] {connectionUrl});
+    }
+
+    @OnDisabled
+    public void close() {
+        getLogger().info("Closing driver");
+        if ( neo4JDriver != null ) {
+            neo4JDriver.close();
+            neo4JDriver = null;
+        }
+    }
+
+    private Map<String, Object> handleInternalNode(Map<String, Object> recordMap) {
+        if (recordMap.size() == 1) {
+            String key = recordMap.keySet().iterator().next();
+            Object value = recordMap.get(key);
+            if (value instanceof InternalNode) {
+                return ((InternalNode)value).asMap();
+            }
+        }
+
+        return recordMap;
+    }
+
+    @Override
+    public Map<String, String> executeQuery(String query, Map<String, Object> parameters, GraphQueryResultCallback handler) {
+        try (Session session = neo4JDriver.session()) {
+            StatementResult result = session.run(query, parameters);
+            long count = 0;
+            while (result.hasNext()) {
+                Record record = result.next();
+                Map<String, Object> asMap = handleInternalNode(record.asMap());
+                handler.process(asMap, result.hasNext());
+                count++;
+            }
+
+            ResultSummary summary = result.summary();
+            SummaryCounters counters = summary.counters();
+
+            Map<String,String> resultAttributes = new HashMap<>();
+            resultAttributes.put(NODES_CREATED,String.valueOf(counters.nodesCreated()));
+            resultAttributes.put(RELATIONS_CREATED,String.valueOf(counters.relationshipsCreated()));
+            resultAttributes.put(LABELS_ADDED,String.valueOf(counters.labelsAdded()));
+            resultAttributes.put(NODES_DELETED,String.valueOf(counters.nodesDeleted()));
+            resultAttributes.put(RELATIONS_DELETED,String.valueOf(counters.relationshipsDeleted()));
+            resultAttributes.put(PROPERTIES_SET, String.valueOf(counters.propertiesSet()));
+            resultAttributes.put(ROWS_RETURNED, String.valueOf(count));
+
+            return resultAttributes;
+        } catch (Exception ex) {
+            getLogger().error("", ex);
+            throw new ProcessException(ex);
```

Comment:
Recommend removing the log and adding details to the exception message:
```suggestion
            throw new ProcessException("Query Failed", ex);
```